packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -530,6 +517,28 @@ export class ConnectionManager implements IConnectionManager {
 			requestedMode = "write";
 		}
 
+		if (this.connection !== undefined || this.pendingConnection !== undefined) {
+			// Connection attempt already completed successfully or is in progress
+			// In general, there should be no issues if the modes do not match:
+			// If at some point it was Ok to connect as "read" (i.e. there were no pending ops we had to track),
+			// then it should be Ok to use "read" connection even if for some reason request came in to connect as "write"
+			// (though that should never happen)
+			// The opposite should be fine as well: we may have had idle "write" connection, and request to reconnect came in,
+			// using default "read" mode.
+			// That all said, let's understand better where such mismatches are coming from.
+			const mode = this.connection?.mode ?? this.pendingConnection?.connectionMode;
+			if (mode !== requestedMode) {
+				this.logger.sendErrorEvent({
+					eventName: "ConnectionModeMismatch",
+					hasConnection: this.connection !== undefined,
+					connectionMode: mode,
```

Comment:
Also, Maybe we no longer need the connection mode mismatch check in the `setupNewSuccessfulConnection` function anymore which logs the ConnectionModeMismatch error event?

Reply:
Interesting, I did not know (or forgot) about "ConnectionModeMismatch" event.
They track slightly different things, though there is some relaionship.
"ConnectionModeMismatch" is more about the cases where service either downgraded our request or upgraded. Both are possible and allowed.
Downgrade is simple - user may not have write permissions to a file, but requested "write" connection. In such case user will get only "read" connection.
Upgrade is allowed by protocol (i.e. user gets more than it asks for), but I do not think any service does it today.

This event is more about what our code wants at different points in time. That said, if service downgraded or upgraded request, then this event may provide somewhat incorrect picture, as it may compare resulting connection mode vs. current request. Resulting connection could be what you get if you request current mode.

I think it's fine to have both events, but I'll downgrade it to generic event - do not want to create scares while we continue to learn more about how system works and where it could have inefficiencies.

Reply:
FYI - I did a bit more due diligence looking through telemetry (Loop & service logs) for any sign that this new workflow would be problematic. I found some interesting things that make system not efficient (posted on our team's chat), but do not see anything that violates this direction.